plan/conflict.go

```diff
@@ -62,6 +64,57 @@ func (s PerResource) ResolveUpdate(current *endpoint.Endpoint, candidates []*end
 	return s.ResolveCreate(candidates)
 }
 
+// ResolveRecordTypes attempts to detect and resolve record type conflicts in desired
+// endpoints for a domain. For eample if the there is more than 1 candidate and at lease one
+// of them is a CNAME. Per [RFC 1034 3.6.2] domains that contain a CNAME can not contain any
+// other record types. The default policy will prefer CNAME record types when a conflict is
+// detected.
+//
+// [RFC 1034 3.6.2]: https://datatracker.ietf.org/doc/html/rfc1034#autoid-15
+func (s PerResource) ResolveRecordTypes(key planKey, row *planTableRow) map[string]*domainEndpoints {
+	if len(row.candidates) <= 1 {
+		return row.records
+	}
+
+	cname := false
+	other := false
+	for _, c := range row.candidates {
+		if c.RecordType == endpoint.RecordTypeCNAME {
+			cname = true
+		} else {
+			other = true
+		}
+
+		if cname && other {
+			break
+		}
+	}
+
+	// conflict was found, remove candiates of non-preferred record types
+	if cname && other {
+		log.Warnf("Domain %s contains conflicting record type candidates, keeping CNAME record", key.dnsName)
+		records := map[string]*domainEndpoints{}
+		for recordType, recs := range row.records {
+			if recordType == endpoint.RecordTypeCNAME {
+				// policy is to prefer the CNAME record type when a conflic is found
```

Comment:
Elsewhere, the behavior is defined that an A or AAAA record is preferred over a CNAME. https://github.com/wbh1/external-dns/blob/f17e24ff8a0ad6223c7c6faeffda683a8f42ef74/endpoint/endpoint.go#L117-L118

That's also been how the conflict resolution has historically functioned, so this could be a major change in behavior. Is that desired / should the new preference be updated elsewhere to be consistent?

Reply:
That's a good point. Perhaps we should treat CNAME as lower priority than other.

Reply:
Updated conflict resolver to prefer A/AAAA over CNAME.